sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala

```diff
@@ -842,3 +699,302 @@ object HyperLogLogPlusPlus {
   )
   // scalastyle:on
 }
+
+abstract class CentralMomentAgg(child: Expression) extends ImperativeAggregate with Serializable {
+
+  /**
+   * The maximum central moment order to be computed
+   */
+  protected def maxMoment: Int
+
+  override def children: Seq[Expression] = Seq(child)
+
+  override def nullable: Boolean = false
+
+  override def dataType: DataType = DoubleType
+
+  // Expected input data type.
+  // TODO: Right now, we replace old aggregate functions (based on AggregateExpression1) to the
+  // new version at planning time (after analysis phase). For now, NullType is added at here
+  // to make it resolved when we have cases like `select avg(null)`.
+  // We can use our analyzer to cast NullType to the default data type of the NumericType once
+  // we remove the old aggregate functions. Then, we will not need NullType at here.
+  override def inputTypes: Seq[AbstractDataType] = Seq(TypeCollection(NumericType, NullType))
+
+  override def aggBufferSchema: StructType = StructType.fromAttributes(aggBufferAttributes)
+
+  /**
+   * The number of central moments to store in the buffer
+   */
+  private[this] val numMoments = 5
+
+  override val aggBufferAttributes: Seq[AttributeReference] = Seq.tabulate(numMoments) { i =>
+    AttributeReference(s"M$i", DoubleType)()
+  }
+
+  /**
+   * Initialize all moments to zero
+   */
+  override def initialize(buffer: MutableRow): Unit = {
+    var aggIndex = 0
+    while (aggIndex < numMoments) {
+      buffer.setDouble(mutableAggBufferOffset + aggIndex, 0.0)
+      aggIndex += 1
+    }
+  }
+
+  /**
+   * Update the central moments buffer.
+   */
+  override def update(buffer: MutableRow, input: InternalRow): Unit = {
+    val v = Cast(child, DoubleType).eval(input)
+    if (v != null) {
+      val updateValue = v match {
+        case d: Double => d
+        case _ => 0.0
+      }
+      val currentM0 = buffer.getDouble(mutableAggBufferOffset)
+      val currentM1 = buffer.getDouble(mutableAggBufferOffset + 1)
+      val currentM2 = buffer.getDouble(mutableAggBufferOffset + 2)
+      val currentM3 = buffer.getDouble(mutableAggBufferOffset + 3)
+      val currentM4 = buffer.getDouble(mutableAggBufferOffset + 4)
+
+      val updateM0 = currentM0 + 1.0
+      val delta = updateValue - currentM1
+      val deltaN = delta / updateM0
+
+      val updateM1 = currentM1 + delta / updateM0
+      val updateM2 = if (maxMoment >= 2) {
+        currentM2 + delta * (delta - deltaN)
+      } else {
+        0.0
+      }
+      val delta2 = delta * delta
```

Comment:
Those computation are not needed if `maxMoment < 3`. So we should move then into the `if (maxMoment >= 3) {` block. Again, changing them to private vars helps here.